vision_evaluation/evaluators.py

```diff
@@ -61,25 +63,149 @@ def add_predictions(self, predictions, targets):
+        """ Add a batch of predictions for evaluation.
+        Args:
+            predictions: the model output array. Shape (N, num_class)
+            targets: the ground truths. Shape (N, num_class) for multi-label (or (N,) for multi-class)
+        """
+
+        assert len(predictions) == len(targets)
+
+        num_samples = len(predictions)
+        target_mat = _targets_to_mat(targets, predictions.shape[1])
+
+        prediction_over_threshold = self.prediction_filter.filter(predictions, 'vec')
+        n_correct_predictions = np.multiply(prediction_over_threshold, target_mat).sum(1)  # shape (N,)
+        n_total = (np.add(prediction_over_threshold, target_mat) >= 1).sum(1)  # shape (N,)
+        n_total[n_total == 0] = 1  # To avoid zero-division. If n_total==0, num should be zero as well.
+        self.sample_accuracy_sum += (n_correct_predictions / n_total).sum()
+        self.num_sample += num_samples
 
     def get_report(self, **kwargs):
-        return {f'top{self.k}_accuracy': float(self.topk_correct_num) / self.total_num if self.total_num else 0.0}
+        return {f'accuracy_{self.prediction_filter.identifier}': float(self.sample_accuracy_sum) / self.num_sample if self.num_sample else 0.0}
 
+    def reset(self):
+        super(ThresholdAccuracyEvaluator, self).reset()
+        self.num_sample = 0
+        self.sample_accuracy_sum = 0
+
+
+class PrecisionEvaluator(Evaluator):
+    """
+    Precision evaluator for both multi-class and multi-label classification
+    """
+
+    def __init__(self, prediction_filter: PredictionFilter):
+        super().__init__()
+        self.prediction_filter = prediction_filter
+
+    def add_predictions(self, predictions, targets):
+        """ Add a batch of predictions for evaluation.
+        Args:
+            predictions: the model output array. Shape (N, num_class)
+            targets: the ground truths. Shape (N, num_class) for multi-label or (N,) for multi-class
+        """
+
+        assert len(predictions) == len(targets)
+
+        num_samples = len(predictions)
+        target_mat = _targets_to_mat(targets, predictions.shape[1])
+
+        filtered_preds = self.prediction_filter.filter(predictions, 'vec')
+        n_correct_predictions = np.multiply(filtered_preds, target_mat).sum(1)  # shape (N,)
+        n_predictions = filtered_preds.sum(1)  # shape (N,)
+        n_predictions[n_predictions == 0] = 1  # To avoid zero-division. If n_predictions==0, num should be zero as well.
+        self.sample_precision_sum += (n_correct_predictions / n_predictions).sum()
+        self.num_sample += num_samples
+
+    def get_report(self, **kwargs):
+        return {f'precision_{self.prediction_filter.identifier}': float(self.sample_precision_sum) / self.num_sample if self.num_sample else 0.0}
+
+    def reset(self):
+        super(PrecisionEvaluator, self).reset()
+        self.num_sample = 0
+        self.sample_precision_sum = 0
+
+
+class RecallEvaluator(Evaluator):
```

Comment:
There seems to be lots of duplicates between PrecisionEvaluator and RecallEvaluator. Can we reduce them?